libraries/core_libs/consensus/src/rewards/block_stats.cpp

```diff
@@ -74,24 +75,43 @@ void BlockStats::processStats(const PeriodData& block) {
   validators_stats_.reserve(std::max(block.dag_blocks.size(), block.previous_block_cert_votes.size()));
   auto block_transactions_hashes_ = toTrxHashesSet(block.transactions);
 
-  for (const auto& dag_block : block.dag_blocks) {
-    const addr_t& dag_block_author = dag_block.getSender();
-    bool has_unique_transactions = false;
-    for (const auto& tx_hash : dag_block.getTrxs()) {
-      // we should also check that we have transactions in pbft block(period data). Because in dag blocks could be
-      // included transaction that was finalized in previous blocks
-      if (!block_transactions_hashes_.contains(tx_hash)) {
-        continue;
+  if (new_dag_rewards_) {
+    uint32_t min_difficulty = UINT32_MAX;
+    for (const auto& dag_block : block.dag_blocks) {
+      if (dag_block.getDifficulty() < min_difficulty) {
+        min_difficulty = dag_block.getDifficulty();
       }
-
-      // returns is transactions was inserted to txs_validators_(its first including into blockchain)
-      if (addTransaction(tx_hash, dag_block_author)) {
-        has_unique_transactions = true;
+    }
+    for (const auto& dag_block : block.dag_blocks) {
+      const addr_t& dag_block_author = dag_block.getSender();
+      if (dag_block.getDifficulty() == min_difficulty) {
+        validators_stats_[dag_block_author].dag_blocks_count_ += 1;
+        total_dag_blocks_count_ += 1;
+      }
+      for (const auto& tx_hash : dag_block.getTrxs()) {
+        addTransaction(tx_hash, dag_block_author);
       }
     }
-    if (has_unique_transactions) {
-      validators_stats_[dag_block_author].dag_blocks_count_ += 1;
-      total_dag_blocks_count_ += 1;
+  } else {
+    for (const auto& dag_block : block.dag_blocks) {
+      const addr_t& dag_block_author = dag_block.getSender();
+      bool has_unique_transactions = false;
+      for (const auto& tx_hash : dag_block.getTrxs()) {
+        // we should also check that we have transactions in pbft block(period data). Because in dag blocks could be
+        // included transaction that was finalized in previous blocks
+        if (!block_transactions_hashes_.contains(tx_hash)) {
+          continue;
+        }
+
+        // returns is transactions was inserted to txs_validators_(its first including into blockchain)
+        if (addTransaction(tx_hash, dag_block_author)) {
+          has_unique_transactions = true;
+        }
+      }
+      if (has_unique_transactions) {
+        validators_stats_[dag_block_author].dag_blocks_count_ += 1;
+        total_dag_blocks_count_ += 1;
+      }
```

Comment:
I think it's better to move it into two separate functions, something like `addVote` below. So it will look something like this:
```cpp
for (const auto& dag_block : block.dag_blocks) {
  if (new_dag_rewards_) {
    addDagBlockDifficulty();
  } else {
    addDagBlock();
  }
}
```

Reply:
Done